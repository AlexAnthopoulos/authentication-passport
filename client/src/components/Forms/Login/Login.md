### Login Form

Visual Example of the Login Form

```js
import { AuthContext } from "../../../contexts/AuthContext";

<AuthContext.Provider value={{}}>
  <Login />
</AuthContext.Provider>;
```