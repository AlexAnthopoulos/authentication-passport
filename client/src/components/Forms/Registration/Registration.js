import React, { useContext } from "react";
import { useForm } from "react-hook-form";
import { Redirect } from "react-router-dom";
import { AuthContext } from "../../../contexts/AuthContext";

export default () => {
  const { redirectLogin, submitRegistrationForm, isAuthenticated } = useContext(
    AuthContext
  );
  const { register, errors, watch, handleSubmit } = useForm();

  const redirectToLogin = () => <Redirect to="/login" />;

  return redirectLogin || isAuthenticated ? (
    redirectToLogin()
  ) : (
    <section data-testid="test-formSection">
      <h2 data-testid="test-formTitle">Register New Account!</h2>
      <p data-testid="test-formDescription">
        Please fill in your information to create your account
      </p>
      <form
        onSubmit={handleSubmit(submitRegistrationForm)}
        data-testid="test-form"
      >
        <div>
          <label htmlFor="firstname" data-testid="test-label">
            First Name
          </label>
          <br />
          <input
            type="text"
            name="firstname"
            placeholder="Enter your first name"
            ref={register({ required: true })}
            data-testid="test-firstname"
          />
          <br />
          {errors.firstname && <span>First Name is required!</span>}
        </div>
        <div>
          <label htmlFor="lastname" data-testid="test-label">
            Last Name
          </label>
          <br />
          <input
            type="text"
            name="lastname"
            placeholder="Enter your last name"
            ref={register({ required: true })}
            data-testid="test-lastname"
          />
          <br />
          {errors.lastname && <span>Last Name is required!</span>}
        </div>
        <div>
          <label htmlFor="email" data-testid="test-label">
            Email
          </label>
          <br />
          <input
            type="email"
            name="email"
            placeholder="Enter your email"
            ref={register({
              required: true,
              pattern: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
            })}
            data-testid="test-email"
          />
          <br />
          {errors.email && errors.email.type === "required" && (
            <span>Email is required!</span>
          )}
          {errors.email && errors.email.type === "pattern" && (
            <span>Email format is wrong!</span>
          )}
        </div>
        <div>
          <label htmlFor="password" data-testid="test-label">
            Password
          </label>
          <br />
          <input
            type="password"
            name="password"
            placeholder="Enter your password"
            ref={register({ required: true, minLength: 6 })}
            data-testid="test-password"
          />
          <br />
          {errors.password &&
            "Password needs to be at least 6 characters long!"}
        </div>
        <div>
          <label htmlFor="verifyPassword" data-testid="test-label">
            Verify Password
          </label>
          <br />
          <input
            type="password"
            name="verifyPassword"
            placeholder="Verify your password"
            ref={register({
              required: true,
              validate: (value) => value === watch("password"),
            })}
            data-testid="test-verifyPassword"
          />
          <br />
          {errors.verifyPassword && "Passwords don't match!"}
        </div>
        <div>
          <button type="submit" data-testid="test-submit">
            Register
          </button>
        </div>
      </form>
    </section>
  );
};
