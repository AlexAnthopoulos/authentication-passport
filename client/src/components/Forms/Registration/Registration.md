### Registration Form

```js
import { AuthContext } from "../../../contexts/AuthContext";

<AuthContext.Provider value={{}}>
  <Registration />
</AuthContext.Provider>;
```
