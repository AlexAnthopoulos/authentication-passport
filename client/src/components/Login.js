import React from "react";
import { useForm } from "react-hook-form";

const Login = () => {
  const { register, errors, handleSubmit } = useForm();

  const submitLogin = (data, e) => {
    e.preventDefault();

    const user = {
      email: data.email,
      password: data.password,
    };

    fetch("/login", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(user),
    })
      .then((response) => response.json())
      .then((results) => console.log(results))
      .catch((loginErr) => console.error(`Log in Fetch sucked: ${loginErr}`));
  };

  return (
    <div>
      <form onSubmit={handleSubmit(submitLogin)}>
        <div>
          <label htmlFor="email">Email</label>
          <br />
          <input
            type="email"
            name="email"
            ref={register({
              required: true,
              pattern: /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/,
            })}
          />
          <br />
          {errors.email && "Email format is wrong!"}
        </div>
        <div>
          <label htmlFor="password">Password</label>
          <br />
          <input
            type="password"
            name="password"
            ref={register({ required: true, minLength: 3 })}
          />
          <br />
          {errors.password &&
            "Password needs to be at least 3 characters long!"}
        </div>
        <button type="submit">Register</button>
      </form>
    </div>
  );
};

export default Login;
