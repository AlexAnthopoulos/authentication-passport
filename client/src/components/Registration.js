import React, { useContext } from "react";
import { useForm } from "react-hook-form";
import { Redirect } from "react-router-dom";
import { AuthContext } from "../contexts/AuthContext";

export default () => {
  const { redirectLogin, submitRegistrationForm, isAuthenticated } = useContext(
    AuthContext
  );
  const { register, errors, watch, handleSubmit } = useForm();

  const redirectToLogin = () => <Redirect to="/login" />;

  return redirectLogin || isAuthenticated ? (
    redirectToLogin()
  ) : (
    <section>
      <h2>Register New Account!</h2>
      <p>Please fill in your information to create your account</p>
      <form onSubmit={handleSubmit(submitRegistrationForm)}>
        <div>
          <label htmlFor="firstname">First Name</label> <br />
          <input
            type="text"
            name="firstname"
            placeholder="Enter your first name"
            ref={register({ required: true })}
          />
          <br />
          {errors.firstname && "First Name is required!"}
        </div>
        <div>
          <label htmlFor="lastname">Last Name</label> <br />
          <input
            type="text"
            name="lastname"
            placeholder="Enter your last name"
            ref={register({ required: true })}
          />
          <br />
          {errors.lastname && "Last Name is required!"}
        </div>
        <div>
          <label htmlFor="email">Email</label> <br />
          <input
            type="email"
            name="email"
            placeholder="Enter your email"
            ref={register({
              required: true,
              pattern: /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/,
            })}
          />
          <br />
          {errors.email && "Email format is wrong!"}
        </div>
        <div>
          <label htmlFor="password">Password</label> <br />
          <input
            type="password"
            name="password"
            placeholder="Enter your password"
            ref={register({ required: true, minLength: 6 })}
          />
          <br />
          {errors.password && "Password needs to be at least 6 characters long!"}
        </div>
        <div>
          <label htmlFor="verifyPassword">Verify Password</label> <br />
          <input
            type="password"
            name="verifyPassword"
            placeholder="Verify your password"
            ref={register({
              required: true,
              validate: (value) => value === watch("password"),
            })}
          />
          <br />
          {errors.verifyPassword && "Passwords don't match!"}
        </div>
        <div>
          <button type="submit">Register</button>
        </div>
      </form>
    </section>
  );
};
