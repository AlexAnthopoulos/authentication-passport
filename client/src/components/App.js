import React, { Component } from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import "./App.css";

import Registration from "./Registration";
import Login from "./Login";
import Profile from "./Profile";
import Navbar from "./Navbar";
import AuthContextProvider from "../contexts/AuthContext";

function App() {
  return (
    <section className="App">
      <Router>
        <Navbar />
        <Switch>
          <AuthContextProvider>
            <Route exact path="/" component={Registration} />
            <Route path="/login" component={Login} />
            <Route path="/profile" component={Profile} />
          </AuthContextProvider>
        </Switch>
      </Router>
    </section>
  );
}

export default App;
