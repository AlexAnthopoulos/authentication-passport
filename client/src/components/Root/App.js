import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import "./App.css";

import Registration from "../Forms/Registration/Registration";
import Login from "../Forms/Login";
import Profile from "../ProtectedComponents/Profile";
import Navbar from "../Navigation/Navbar";
import AuthContextProvider from "../../contexts/AuthContext";
import PrivateRoute from "../ProtectedRoutes/PrivateRoute";

function App() {
  return (
    <section className="App">
      <AuthContextProvider>
        <Router>
          <Navbar />
          <Switch>
            <Route exact path="/" component={Registration} />
            <Route path="/login" component={Login} />
            <PrivateRoute path="/profile" component={Profile} />
          </Switch>
        </Router>
      </AuthContextProvider>
    </section>
  );
}

export default App;
