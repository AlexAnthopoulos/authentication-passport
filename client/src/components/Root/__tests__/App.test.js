import React from "react";
import App from "../App";
import { Route } from "react-router";
import { shallow } from "enzyme";
import Registration from "../../Forms/Registration";
import Login from "../../Forms/Login";

const shallowComponent = () => shallow(<App />); // Real unit test (isolation, no children render)
const wrapper = shallowComponent();

describe("Unit tests for App component", () => {
  test("shallow render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test(".App styling class exists", () => {
    expect(wrapper.find(".App").hasClass("App")).toBe(true);
  });

  test("public routes rendering the correct components", () => {
    const routesEndpoints = wrapper
      .find(Route)
      .reduce((routesEndpoints, route) => {
        const routesProps = route.props(); // Gets the routesProps from the routes
        routesEndpoints[routesProps.path] = routesProps.component;
        return routesEndpoints;
      }, {});

    expect(routesEndpoints["/"]).toBe(Registration);
    expect(routesEndpoints["/login"]).toBe(Login);
  });
});
