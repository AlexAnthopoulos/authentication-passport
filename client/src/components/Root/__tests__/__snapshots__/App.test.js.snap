// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow render 1`] = `
<section
  className="App"
>
  <AuthContextProvider>
    <BrowserRouter>
      <Navbar />
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          path="/login"
        />
        <_default
          component={[Function]}
          path="/profile"
        />
      </Switch>
    </BrowserRouter>
  </AuthContextProvider>
</section>
`;
